File: CS_Algorithms/ADA_Algorithms/Greedy_Algorithms/Python/prims_algorithm.py
from typing import List, Tuple, Any, Dict 
import heapq 

class Graph:
    def __init__(self, vertices: int):
        self.vertices = vertices 
        self.graph = [[] for _ in range(vertices)]
        
    def add_edge_in_graph(self, src: int, dest: int, wt: int):
        self.graph[src].append((dest, wt))
        self.graph[dest].append((src, wt))
    
class Prim:
    def __init__(self,graph:Graph):
        self.graph=graph 
    
    def find_mst(self):
        visited = [False]*self.graph.vertices
        mst=[]
        cost = 0 
        min_heap=[]
        heapq.heappush(min_heap,(0,0,-1)) # (wt,v2,v1)
        while min_heap:
            wt,vertex,parent = heapq.heappop(min_heap)
            if visited[vertex]:
                continue
            visited[vertex] = True  
            cost += wt 
            if parent != -1:
                mst.append((parent,vertex,wt))
            for neighbor,wt in self.graph.graph[vertex]:
                if not visited[neighbor]:
                    heapq.heappush(min_heap,(wt,neighbor,vertex))
        return mst,cost

def main():
    g = Graph(5)
    g.add_edge_in_graph(0,1,2)
    g.add_edge_in_graph(0,3,4)
    g.add_edge_in_graph(1,2,5)
    g.add_edge_in_graph(2,4,1)
    g.add_edge_in_graph(3,4,2)
    g.add_edge_in_graph(2,3,6)
    p = Prim(g)
    mst, cost = p.find_mst()
    for src,dest,wt in mst:
        print(f"{src} - {dest} => {wt}")
    print(cost)

if __name__ == "__main__":
    main()

File: CS_Algorithms/ADA_Algorithms/Greedy_Algorithms/C++/prims_algorithm.cpp
#include<iostream>
#include<vector>
#include<queue>
using namespace std ; 

class Graph{
    int vertices; vector<vector<pair<int,int>>>adjList; 
public: 
    Graph(int v){ vertices=v ; adjList.resize(v); }
    void addEdgeinGraph(int src, int des, int wt){
        adjList[src].push_back({wt,des});
        adjList[des].push_back({wt,src});  
    }
    int getVertices() const{ return vertices; }
    const vector<pair<int,int>>& getNeighbors(int vertex) const{ return adjList[vertex];}
};

class Prim{
    Graph& g; 
public:
    explicit Prim(Graph&graph): g(graph){}
    void findMST(){
        vector<bool>visited(g.getVertices(),false);
        using Edge = pair<int,pair<int,int>>; 
        priority_queue<Edge,vector<Edge>,greater<Edge>>min_heap;
        min_heap.push({0,{0,-1}});
        int cost = 0 ; 
        while(!min_heap.empty()){
            Edge curr = min_heap.top(); min_heap.pop(); 
            int wt = curr.first; int vertex = curr.second.first; int par = curr.second.second;
            if(visited[vertex])continue; 
            visited[vertex]=true; cost+= wt; 
            if(par!=-1)cout<<par<<" "<<vertex<<" "<<wt<<endl;
            for(const auto& neighbor : g.getNeighbors(vertex)){
                int w = neighbor.first; int v1 = neighbor.second; 
                if(!visited[v1]){min_heap.push({w,{v1,vertex}});}
            }
        }
        cout<<cost<<endl ; 
    }
};

int main(){
    Graph g(5); 
    g.addEdgeinGraph(0,1,2); g.addEdgeinGraph(0,3,4);
    g.addEdgeinGraph(1,2,5); g.addEdgeinGraph(2,4,1); 
    g.addEdgeinGraph(3,4,2); g.addEdgeinGraph(2,3,6); 

    Prim p(g); 
    p.findMST(); return 0 ; 
}